### PR TITLE
[IOTDB-863]: add a switch to drop ouf-of-order data that is recognized by IoTDB

### PR DESCRIPTION
**Pull request by haimeiguo**
Users may not care about out of order data.
So I add a configuration parameter (enable_discard_out_of_order_data) to allow iotdb to discard the out-of-order data

jira: https://issues.apache.org/jira/projects/IOTDB/issues/IOTDB-863?filter=allopenissues

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java [hunk 1]
```diff
@@ -808,8 +815,11 @@ public void insertTablet(InsertTabletPlan insertTabletPlan) throws BatchInsertio
           // judge if we should insert sequence
           if (!isSequence && time > lastFlushTime) {
```
Comment (mychaow) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java:816]: so, the first row of one timepartition always insert into unsequence memtable?

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java [hunk 2]
```diff
@@ -144,6 +144,8 @@
 
   private volatile boolean readOnly = false;
 
+  private boolean enable_discard_out_of_order_data = false;
```
Comment (samperson1997) [inline server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java:147]:
```suggestion
  private boolean enableDiscardOutOfOrderData = false;
```

**Review by samperson1997 — COMMENTED** [review]
Hi, mind the naming rule : )

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java (hunk 2 above)
Comment (haimeiguo) [reply server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java]: edited! thank you

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java [hunk 3]
```diff
@@ -733,10 +733,15 @@ public void insert(InsertRowPlan insertRowPlan) throws WriteProcessException {
       partitionLatestFlushedTimeForEachDevice
           .computeIfAbsent(timePartitionId, id -> new HashMap<>());
 
+      boolean isSequence = insertRowPlan.getTime() > partitionLatestFlushedTimeForEachDevice.get(timePartitionId)
```
[the remark this answers is not in the document]
Comment (haimeiguo) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java:735]: edited! thank you

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java (hunk 2 above)
Comment (samperson1997) [reply server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java]: My pleasure ; )

**Review by mychaow — APPROVED** [review]

File: server/src/test/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessorTest.java [hunk 4]
```diff
@@ -273,6 +274,112 @@ public void testSeqAndUnSeqSyncClose()
     }
   }
 
+  @Test
+  public void testEnableDiscardOutOfOrderDataForInsertRowPlan()
+      throws WriteProcessException, QueryProcessException, IllegalPathException {
+    IoTDBConfig config = IoTDBDescriptor.getInstance().getConfig();
+    config.setEnableDiscardOutOfOrderData(true);
```
Comment (qiaojialin) [inline server/src/test/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessorTest.java:281]: remember to reset this to its default value in the end of the test

We usually do this as follows:

boolean defaultValue = config.isEnableDiscard...;
config.setEnableDiscard(true);
do some test
config.setEnableDiscard(defaultValue);

**Review by qiaojialin — COMMENTED** [review]
Good work! @SilverNarcissus  have a look.

File: server/src/test/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessorTest.java [hunk 5]
```diff
@@ -273,6 +274,112 @@ public void testSeqAndUnSeqSyncClose()
...
+    insertTabletPlan2.setDeviceMNode(deviceMNode);
+
+    processor.insertTablet(insertTabletPlan2);
+    processor.asyncCloseAllWorkingTsFileProcessors();
+    processor.syncCloseAllWorkingTsFileProcessors();
+
+    QueryDataSource queryDataSource = processor.query(new PartialPath(deviceId), measurementId, context,
+        null, null);
+
+    Assert.assertEquals(2, queryDataSource.getSeqResources().size());
```
Comment (SilverNarcissus) [inline server/src/test/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessorTest.java:376]: You should also add some flush operation to test your code

**Review by SilverNarcissus — CHANGES_REQUESTED** [review]
Nice work!  We can control our unsequence data now!   But there are some issues and incomplete test, please have a look~

File: server/src/test/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessorTest.java [hunk 6]
```diff
@@ -26,6 +26,7 @@
 import java.util.Collections;
```
Comment (SilverNarcissus) [inline server/src/test/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessorTest.java:26]: use this line to change the time partition and add some test over different time partition. ```IoTDBDescriptor.getInstance().getConfig().setPartitionInterval(100);```

do not forget to set it back to original value. You can refer to ```IoTDBMultiSeriesIT.java``` for this. Thank you very much XD

**Review by SilverNarcissus — CHANGES_REQUESTED** [review]
Only one issue, please have a look~

File: server/src/test/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessorTest.java (hunk 6 above)
Comment (haimeiguo) [reply server/src/test/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessorTest.java:26]: Thank you for your review! I added test cases for time partition = 100, 500 and 1000.

File: server/src/test/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessorTest.java [hunk 7]
```diff
@@ -267,6 +269,279 @@ public void testSeqAndUnSeqSyncClose()
...
+  }
+
+  @Test
+  public void testEnableDiscardOutOfOrderDataForInsertTablet3()
+      throws QueryProcessException, IllegalPathException, IOException {
+    IoTDBConfig config = IoTDBDescriptor.getInstance().getConfig();
+    boolean defaultValue = config.isEnableDiscardOutOfOrderData();
+    config.setEnableDiscardOutOfOrderData(true);
+    long defaultTimePartition = IoTDBDescriptor.getInstance().getConfig().getPartitionInterval();
+    config.setPartitionInterval(1000);
```
Comment (SilverNarcissus) [inline server/src/test/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessorTest.java:475]: Sure, you set time partition. But you insert time range only from 0 -150 which never cross the partition interval. By the way, you should also enable the partition like this:
```
config.setEnablePartition(true);
```
and do not forget:
```
config.setEnablePartition(false);
```

Thank you very much~

**Review by SilverNarcissus — CHANGES_REQUESTED** [review]

File: server/src/test/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessorTest.java (hunk 7 above)
Comment (qiaojialin) [reply server/src/test/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessorTest.java:475]: boolean defaultEnablePartition = config.getEnablePartition...

**Review by SilverNarcissus — APPROVED** [review]
Nice work! Look good to me~

**Pull request merged**